design/baremetal-operator/bmh-v1beta1.md

```diff
@@ -0,0 +1,140 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# BareMetalHost API v1beta1
+
+## Status
+
+One of: implementable
+
+## Summary
+
+This design proposes making certain backward incompatible changes to the
+BareMetalHost API and publishing the result as version `v1beta1`.
+
+## Motivation
+
+The BareMetalHost API is currently stuck at `v1alpha1`. Meanwhile, several
+incompatible changes are needed based on the recent development both in Metal3
+itself and in Ironic.
+
+For example, Ironic is planning a complete removal of MD5 support for
+checksums.
+
+### Goals
+
+- Clean up issues in the BareMetalHost API that have accumulated during
+  the usage of `v1alpha1`.
+
+- Make the API version reflect the actual support status of the API (see
+  [alternatives](#Alternatives) for details).
+
+### Non-Goals
+
+- Stabilize the BareMetalHost API completely (as `v1`).
+
+## Proposal
+
+Copy the current API types definition into a new package `v1beta1` with some
+modifications.
+
+## Design Details
+
+The following addition should be done *before* the version is created:
+
+- Add a `Spec.Architecture` field to override architecture in the profile.
+  If not provided, BMO will use its own architecture as the default.
+
+The following changes will be done to the new version:
+
+- Remove MD5 support from `Image.ChecksumType`. When no `ChecksumType` is
+  provided, do not populate the `os_image_hash_algo` field and let Ironic
+  guess the algorithm.
+
+  The same change is applied to CAPM3.
+
+- In `RootDeviceHints`:
+
+  Rename `DeviceName` to `DevicePath` to better reflect what it actually is
+  (especially after introducing support for `/dev/disk/by-path`).
+
+  Rename `Model` and `Vendor` to `ModelContains` and `VendorContains` to
+  reflect the fact that these fields use the `<in>` operator.
+
+- Remove `HardwareProfile` and don't do any matching.
+
+- Remove `HardwareDetails` from the `Status` in favour of the already
+  introduced `HardwareData` CR.
+
+- Rename `online` to `poweredOn`.
+
+  The word *online* is ambiguous here: it may be understood to refer to
+  provisioning status. Furthermore, the corresponding field in the `Status`
+  is already called `poweredOn`.
+
+### Implementation Details/Notes/Constraints
+
+Both `v1alpha1` and the proposed `v1beta1` resources will be possible
+to handle with the same controller code since the differences are mostly
+cosmetic.
+
+### Risks and Mitigations
+
+None
+
+### Work Items
+
+- Add `Architecture`
+- Refactor the BMO code to avoid referring directly to the current version.
+- Copy type definitions and modify them.
+- Create a conversion webhook to change between versions.
+- Communicate the change to the community.
+
+### Dependencies
+
+None
+
+### Test Plan
+
```

Comment:
I was thinking about adding an implementation schedule of checklist.
IMO the implementation process has to already incorporate the branching that we have talked about during MTM.

I would like to have this schedule:
- Implement the aplhav1 related changes e.g. spec.Architecture
- Branch out to alphav1 branch after there are nothing to do to the old api
- Release a patch version of BMO from the alpha v1 branch
- Continue  implementation of v1Beta1 specific features on main branch 
- Branch out v1Beta1
- release v1Beta1 as major 1.0 release of BMO
--- future:
- start planning the scope of v1 in order to have some idea about the future

Reply:
more about how the branches would be supported here: https://github.com/metal3-io/baremetal-operator/issues/1283

Reply:
Isn't "Work items" the checklist?

I'm not sure we need to tie supported branches with supported API versions. Not saying it's a bad idea, but in theory we can just introduce a new version and only branch out before removing the old version.

Reply:
Yep the work tickets are fine, I see that my comment was badly written I actually didn't mean follow the API versions with the branches, I would like to follow this metal3-io/baremetal-operator#1283 branching out at the release of every minor and major version.